fdbserver/workloads/ValidateStorage.actor.cpp

```diff
@@ -0,0 +1,248 @@
+
+		TraceEvent("TestValueWritten");
+
+		wait(self->validateData(self, cx, KeyRangeRef("TestKeyA"_sr, "TestKeyF"_sr)));
+		TraceEvent("TestValueVerified");
+
+		loop {
+			try {
+				UID auditId = wait(auditStorage(cx->getConnectionRecord(), allKeys, AuditType::ValidateHA));
+				TraceEvent("TestValidateEnd").detail("AuditID", auditId);
+				break;
+			} catch (Error& e) {
+				TraceEvent("AuditStorageError").errorUnsuppressed(e);
+				wait(delay(1));
+			}
+		}
+
+		return Void();
+	}
+
+	ACTOR Future<Version> populateData(ValidateStorage* self, Database cx, std::map<Key, Value>* kvs) {
+		state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(cx);
+		state Version version;
+		loop {
+			state UID debugID = deterministicRandom()->randomUniqueID();
+			try {
+				tr->debugTransaction(debugID);
+				for (const auto& [key, value] : *kvs) {
+					tr->set(key, value);
+				}
+				wait(tr->commit());
+				version = tr->getCommittedVersion();
+				break;
+			} catch (Error& e) {
+				TraceEvent("TestCommitError").errorUnsuppressed(e);
+				wait(tr->onError(e));
+			}
+		}
+
+		TraceEvent("PopulateTestDataDone")
+		    .detail("CommitVersion", tr->getCommittedVersion())
+		    .detail("DebugID", debugID);
+
+		return version;
+	}
+
+	ACTOR Future<Void> validateData(ValidateStorage* self, Database cx, KeyRange range) {
+		TraceEvent("TestValidateStorageBegin").detail("Range", range);
+		state Transaction tr(cx);
+		tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+		tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+
+		loop {
+			try {
+				state RangeResult shards =
+				    wait(krmGetRanges(&tr, keyServersPrefix, range, CLIENT_KNOBS->TOO_MANY, CLIENT_KNOBS->TOO_MANY));
+				ASSERT(!shards.empty() && !shards.more);
+
+				state RangeResult UIDtoTagMap = wait(tr.getRange(serverTagKeys, CLIENT_KNOBS->TOO_MANY));
+				ASSERT(!UIDtoTagMap.more && UIDtoTagMap.size() < CLIENT_KNOBS->TOO_MANY);
+
+				state int i = 0;
+				for (i = 0; i < shards.size() - 1; ++i) {
+					std::vector<UID> src;
+					std::vector<UID> dest;
+					UID srcId, destId;
+					decodeKeyServersValue(UIDtoTagMap, shards[i].value, src, dest, srcId, destId);
+
+					const int idx = deterministicRandom()->randomInt(0, src.size());
+					Optional<Value> serverListValue = wait(tr.get(serverListKeyFor(src[idx])));
+					ASSERT(serverListValue.present());
+					const StorageServerInterface ssi = decodeServerListValue(serverListValue.get());
+					TraceEvent("TestValidateStorageSendingRequest")
+					    .detail("Range", range)
+					    .detail("StorageServer", ssi.toString());
+					AuditStorageRequest req(deterministicRandom()->randomUniqueID(),
+					                        KeyRangeRef(shards[i].key, shards[i + 1].key),
+					                        AuditType::ValidateHA);
+					Optional<AuditStorageState> vResult =
+					    wait(timeout<AuditStorageState>(ssi.auditStorage.getReply(req), 5));
+					if (!vResult.present()) {
```

Comment:
Is `AuditStorageRequest` checked synchronously?

Reply:
Yes, it wait for the storage servers's response.